### PR TITLE
syncer(dm): cache some metrics looking up in hot path

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref #5183 

### What is changed and how it works?

- Syncer holds a metrics cache, so some of metrics looking up can be omitted
- Only check checkpoint regression at debug level
- No need to add timeout when GetEvent, it's just reading a channel
- only update location recorder at debug level

### Check List 

Tests 

 - pass original tests
 - **waiting for benchbot test**

Code changes

 - Has exported variable/fields change

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Improve syncer's performance
```

### CODE REVIEW

**Comment by lance6716** [comment]
cc @D3Hunter

File: dm/syncer/syncer.go [hunk 1]
```diff
@@ -1684,7 +1686,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 	s.tctx.L().Info("replicate binlog from checkpoint", zap.Stringer("checkpoint", lastLocation))
 
 	if s.streamerController.IsClosed() {
-		s.locations.reset(lastLocation)
+		if ce := s.tctx.L().Check(zap.DebugLevel, "update locations"); ce != nil {
+			s.locations.reset(lastLocation)
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:1690]: branch(not log itself) depends on log level make it hard to understand, and this is not frequent called, prefer to remove it completely

File: dm/syncer/syncer.go [hunk 2]
```diff
@@ -1781,7 +1785,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		// if suffix>0, we are replacing error
 		s.isReplacingOrInjectingErr = currentLocation.Suffix != 0
 
-		s.locations.reset(currentLocation)
+		if ce := s.tctx.L().Check(zap.DebugLevel, "update locations"); ce != nil {
+			s.locations.reset(currentLocation)
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:1789]: here too

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -1847,7 +1853,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			currentLocation = shardingReSync.currLocation
 			// if suffix>0, we are replacing error
 			s.isReplacingOrInjectingErr = currentLocation.Suffix != 0
-			s.locations.reset(shardingReSync.currLocation)
+			if ce := s.tctx.L().Check(zap.DebugLevel, "update locations"); ce != nil {
+				s.locations.reset(shardingReSync.currLocation)
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:1857]: if it's not needed, maybe just remove it. doesn't help to locate problems if anything happened.

we can add it back later if necessiary

File: dm/syncer/syncer.go [hunk 4]
```diff
@@ -2033,7 +2041,10 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			if currentLocation.Suffix > 0 && e.Header.EventSize > 0 {
 				currentLocation.Suffix = 0
 				s.isReplacingOrInjectingErr = false
-				s.locations.reset(currentLocation)
+				if ce := s.tctx.L().Check(zap.DebugLevel, "update locations"); ce != nil {
+					s.locations.reset(currentLocation)
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:2045]: here too

File: dm/syncer/syncer.go [hunk 5]
```diff
@@ -3770,7 +3781,9 @@ func (s *Syncer) getEvent(tctx *tcontext.Context, startLocation binlog.Location)
 
 	e, err := s.streamerController.GetEvent(tctx)
 	if err == nil {
-		s.locations.update(e)
+		if ce := s.tctx.L().Check(zap.DebugLevel, "update locations"); ce != nil {
+			s.locations.update(e)
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:3785]: doittoo

File: dm/syncer/checkpoint.go [hunk 6]
```diff
@@ -100,9 +100,11 @@ func (b *binlogPoint) save(location binlog.Location, ti *model.TableInfo) error
 	b.Lock()
 	defer b.Unlock()
 
-	if binlog.CompareLocation(location, b.savedPoint.location, b.enableGTID) < 0 {
-		// support to save equal location, but not older location
-		return terror.ErrCheckpointSaveInvalidPos.Generate(location, b.savedPoint.location)
+	if ce := log.L().Check(zap.DebugLevel, "check binlog point regression"); ce != nil {
```
Comment (D3Hunter) [inline dm/syncer/checkpoint.go:103]: please add some comment to explain why code other than log is guarded by log level.

**Comment by D3Hunter** [comment]
maybe add a pr to change relay log too

**Comment by D3Hunter** [comment]
> time to consume 1G binlog (generated by sysbench, GTID on, no relay, no data matching allow-list)
> 
> before [e5eeff4](https://github.com/pingcap/tiflow/commit/e5eeff4d67af744285b1dc4997688070443acfad) : 52s after [e5eeff4](https://github.com/pingcap/tiflow/commit/e5eeff4d67af744285b1dc4997688070443acfad) : 37s after [80bb492](https://github.com/pingcap/tiflow/commit/80bb49219ae26dbf63967b5b15dc868543df52af) : 18s
> 
> cc @GMHDBJD

in my bench(sysbench generated data and using tispace), binlog process speed boost from 5MB/s to 10MB/s, good job.

**Comment by lance6716** [comment]
waiting for benchbot to iterate more quickly. please help us! @XuJianxu

**Comment by lance6716** [comment]
after a recent benchmark, these function hotspots have been resolved

**Pull request closed without merging**